File: sdk/node/src/production-signing.ts
import crypto from "crypto";

export const PRODUCTION_SIGNATURE_DOMAIN = Buffer.from(
  "CAP-PRODUCTION-SIGNATURE-V1\0",
  "utf8",
);

export class ProductionWireError extends Error {}

export interface ExtractedProductionWire {
  unsigned: Uint8Array;
  signature: Uint8Array;
}

export function extractProductionSignature(raw: Uint8Array): ExtractedProductionWire {
  const source = Buffer.from(raw);
  const unsigned: Buffer[] = [];
  let signature: Buffer | undefined;
  let offset = 0;
  while (offset < source.length) {
    const start = offset;
    const tag = readVarint(source, offset);
    offset = tag.end;
    const fieldNumber = Number(tag.value >> 3n);
    const wireType = Number(tag.value & 7n);
    const valueEnd = skipValue(source, offset, wireType);
    if (fieldNumber !== 14) {
      unsigned.push(source.subarray(start, valueEnd));
    } else {
      if (signature || wireType !== 2) {
        throw new ProductionWireError("duplicate or malformed signature field");
      }
      const length = readVarint(source, offset);
      const end = length.end + Number(length.value);
      if (end !== valueEnd || end === length.end) {
        throw new ProductionWireError("malformed signature");
      }
      signature = source.subarray(length.end, end);
    }
    offset = valueEnd;
  }
  if (!signature) throw new ProductionWireError("missing signature");
  return { unsigned: Buffer.concat(unsigned), signature };
}

export function productionPreimageDigest(unsigned: Uint8Array): Uint8Array {
  return crypto
    .createHash("sha256")
    .update(PRODUCTION_SIGNATURE_DOMAIN)
    .update(unsigned)
    .digest();
}

export function verifyProductionSignature(raw: Uint8Array, publicKeyPem: string): Uint8Array {
  const extracted = extractProductionSignature(raw);
  const verifier = crypto.createVerify("sha256");
  verifier.update(PRODUCTION_SIGNATURE_DOMAIN);
  verifier.update(extracted.unsigned);
  verifier.end();
  if (!verifier.verify(publicKeyPem, extracted.signature)) {
    throw new ProductionWireError("invalid signature");
  }
  return extracted.unsigned;
}

export const PRODUCTION_PROFILE_VERSION = "cap-production-v1";
export const PRODUCTION_ALGORITHM = "ECDSA-P256-SHA256";

/** Trust configuration for {@link verifyProductionPacket}. Keys are keyed by
 * signature_metadata.key_id (never trusted from the packet itself). */
export interface ProductionTrustStore {
  audience: string;
  publicKeys: Record<string, string>;
  tenant?: string;
  sender?: string;
}

function timestampMillis(value: any): number {
  if (!value) return 0;
  const seconds = typeof value.seconds === "object" ? Number(value.seconds.toString()) : Number(value.seconds ?? 0);
  const nanos = Number(value.nanos ?? 0);
  return seconds * 1000 + Math.floor(nanos / 1e6);
}

function validateMetadataShape(metadata: any): void {
  if (!metadata) {
    throw new ProductionWireError("missing signature metadata");
  }
  const messageId: Uint8Array = metadata.messageId ?? new Uint8Array(0);
  if (
    metadata.profileVersion !== PRODUCTION_PROFILE_VERSION ||
    metadata.algorithm !== PRODUCTION_ALGORITHM ||
    messageId.length !== 16 ||
    !metadata.audience ||
    !metadata.keyId ||
    !metadata.expiresAt
  ) {
    throw new ProductionWireError("invalid signature metadata");
  }
}

/**
 * Verify exact received wire bytes (never a re-serialized object) and
 * return the decoded, verified BusPacket. busPacketType is the protobufjs
 * Type used to decode the (already-extracted) unsigned bytes, matching
 * this runtime's existing reflection-based decode convention.
 */
export function verifyProductionPacket(raw: Uint8Array, busPacketType: { decode: (buf: Uint8Array) => any }, trust: ProductionTrustStore): any {
  const extracted = extractProductionSignature(raw);
  const packet = busPacketType.decode(extracted.unsigned);
  const metadata = packet.signatureMetadata;
  validateMetadataShape(metadata);
  if (trust.audience && metadata.audience !== trust.audience) {
    throw new ProductionWireError("audience mismatch");
  }
  if (timestampMillis(metadata.expiresAt) <= Date.now()) {
    throw new ProductionWireError("signature expired");
  }
  if (trust.tenant && packet.identity?.tenantId !== trust.tenant) {
    throw new ProductionWireError("tenant mismatch");
  }
  if (trust.sender && packet.senderId !== trust.sender) {
    throw new ProductionWireError("sender mismatch");
  }
  const publicKeyPem = trust.publicKeys[metadata.keyId];
  if (!publicKeyPem) {
    throw new ProductionWireError("unknown key id");
  }
  const verifier = crypto.createVerify("sha256");
  verifier.update(PRODUCTION_SIGNATURE_DOMAIN);
  verifier.update(Buffer.from(extracted.unsigned));
  verifier.end();
  if (!verifier.verify(publicKeyPem, Buffer.from(extracted.signature))) {
    throw new ProductionWireError("invalid signature");
  }
  return packet;
}

interface VarintResult {
  value: bigint;
  end: number;
}

function readVarint(raw: Buffer, start: number): VarintResult {
  let value = 0n;
  for (let index = 0; index < 10; index += 1) {
    const offset = start + index;
    if (offset >= raw.length) throw new ProductionWireError("truncated varint");
    const byte = raw[offset];
    value |= BigInt(byte & 0x7f) << BigInt(index * 7);
    if (byte < 0x80) {
      const encoded = encodeVarint(value);
      if (!encoded.equals(raw.subarray(start, offset + 1))) {
        throw new ProductionWireError("non-minimal varint");
      }
      return { value, end: offset + 1 };
    }
  }
  throw new ProductionWireError("oversize varint");
}

function encodeVarint(input: bigint): Buffer {
  const bytes: number[] = [];
  let value = input;
  while (value >= 0x80n) {
    bytes.push(Number(value & 0x7fn) | 0x80);
    value >>= 7n;
  }
  bytes.push(Number(value));
  return Buffer.from(bytes);
}

function skipValue(raw: Buffer, offset: number, wireType: number): number {
  let end: number;
  if (wireType === 0) return readVarint(raw, offset).end;
  if (wireType === 1) end = offset + 8;
  else if (wireType === 2) {
    const length = readVarint(raw, offset);
    end = length.end + Number(length.value);
  } else if (wireType === 5) end = offset + 4;
  else throw new ProductionWireError("unsupported wire type");
  if (end > raw.length) throw new ProductionWireError("truncated field");
  return end;
}

File: sdk/go/production_signing.go
package capsdk

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"strings"
	"time"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	"google.golang.org/protobuf/encoding/protowire"
	"google.golang.org/protobuf/proto"
)

const (
	ProductionProfileVersion     = "cap-production-v1"
	ProductionAlgorithm          = "ECDSA-P256-SHA256"
	DefaultProductionMaxLifetime = 5 * time.Minute
	MaximumProductionClockSkew   = time.Minute
	DefaultProductionMaxRawBytes = 1 << 20
)

var productionDomain = []byte("CAP-PRODUCTION-SIGNATURE-V1\x00")

var (
	ErrMissingSignatureMetadata = errors.New("capsdk: missing signature metadata")
	ErrSignatureExpired         = errors.New("capsdk: signature expired")
	ErrAudienceMismatch         = errors.New("capsdk: signature audience mismatch")
	ErrUnknownKeyID             = errors.New("capsdk: unknown key id")
	ErrDuplicateSignatureField  = errors.New("capsdk: duplicate signature field")
	ErrMalformedProductionWire  = errors.New("capsdk: malformed production wire")
	ErrUnsupportedProductionKey = errors.New("capsdk: production signatures require ECDSA P-256")
	ErrProductionTrustRequired  = errors.New("capsdk: production verifier requires audience, tenant, sender, and local keys")
)

type ProductionTrustStore struct {
	Audience    string
	Tenant      string
	Sender      string
	PublicKeys  map[string]*ecdsa.PublicKey
	ResolveKey  func(tenant, sender, keyID string) (*ecdsa.PublicKey, error)
	Now         func() time.Time
	MaxLifetime time.Duration
	ClockSkew   time.Duration
}

func SignProductionPacket(packet *agentv1.BusPacket, key *ecdsa.PrivateKey) ([]byte, error) {
	if packet == nil || key == nil {
		return nil, errors.New("capsdk: packet and private key are required")
	}
	if !validProductionPublicKey(&key.PublicKey) || key.D == nil {
		return nil, ErrUnsupportedProductionKey
	}
	if err := validateMetadataForSigning(packet.GetSignatureMetadata(), time.Now()); err != nil {
		return nil, err
	}
	if err := validateNoProductionUnknowns(packet.ProtoReflect()); err != nil {
		return nil, err
	}
	unsignedPacket := proto.Clone(packet).(*agentv1.BusPacket)
	unsignedPacket.Signature = nil
	unsigned, err := proto.Marshal(unsignedPacket)
	if err != nil {
		return nil, fmt.Errorf("marshal unsigned production packet: %w", err)
	}
	if err := validateUnsignedProductionWire(unsigned); err != nil {
		return nil, err
	}
	digest := productionDigest(unsigned)
	signature, err := ecdsa.SignASN1(rand.Reader, key, digest[:])
	if err != nil {
		return nil, fmt.Errorf("sign production packet: %w", err)
	}
	raw := appendSignatureField(unsigned, signature)
	if len(raw) > DefaultProductionMaxRawBytes {
		return nil, ErrMalformedProductionWire
	}
	return raw, nil
}

func VerifyProductionPacket(raw []byte, trust ProductionTrustStore) (*agentv1.BusPacket, error) {
	if err := validateProductionTrust(trust); err != nil {
		return nil, err
	}
	unsigned, signature, err := extractSignatureField(raw)
	if err != nil {
		return nil, err
	}
	packet := &agentv1.BusPacket{}
	if err := proto.Unmarshal(unsigned, packet); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrMalformedProductionWire, err)
	}
	if err := validateNoProductionUnknowns(packet.ProtoReflect()); err != nil {
		return nil, err
	}
	metadata := packet.GetSignatureMetadata()
	if err := validateMetadata(metadata, trust); err != nil {
		return nil, err
	}
	key, err := resolveProductionKey(packet, metadata.GetKeyId(), trust)
	if err != nil {
		return nil, err
	}
	digest := productionDigest(unsigned)
	if !ecdsa.VerifyASN1(key, digest[:], signature) {
		return nil, ErrInvalidSignature
	}
	packet.Signature = append([]byte(nil), signature...)
	return packet, nil
}

func validateProductionTrust(trust ProductionTrustStore) error {
	if !validProductionAuthority(trust.Audience) || !validProductionAuthority(trust.Tenant) ||
		!validProductionAuthority(trust.Sender) ||
		(trust.ResolveKey == nil && len(trust.PublicKeys) == 0) {
		return ErrProductionTrustRequired
	}
	return nil
}

func validProductionAuthority(value string) bool {
	return value != "" && strings.TrimSpace(value) == value
}

// ProductionSignedBodyDigest returns SHA-256 over the exact unsigned wire
// bytes covered by a CAP-PRODUCTION signature. Replay stores use this digest
// so a freshly generated ECDSA signature over the same body remains an
// idempotent redelivery rather than a false conflict.
func ProductionSignedBodyDigest(raw []byte) ([32]byte, error) {
	unsigned, _, err := extractSignatureField(raw)
	if err != nil {
		return [32]byte{}, err
	}
	return sha256.Sum256(unsigned), nil
}

func productionDigest(unsigned []byte) [32]byte {
	input := make([]byte, 0, len(productionDomain)+len(unsigned))
	input = append(input, productionDomain...)
	input = append(input, unsigned...)
	return sha256.Sum256(input)
}

func appendSignatureField(unsigned, signature []byte) []byte {
	result := append([]byte(nil), unsigned...)
	result = protowire.AppendTag(result, 14, protowire.BytesType)
	return protowire.AppendBytes(result, signature)
}

func validateMetadataShape(metadata *agentv1.SignatureMetadata) error {
	if metadata == nil {
		return ErrMissingSignatureMetadata
	}
	expiresAt := metadata.GetExpiresAt()
	if metadata.GetProfileVersion() != ProductionProfileVersion || metadata.GetAlgorithm() != ProductionAlgorithm ||
		len(metadata.GetMessageId()) != 16 || metadata.GetAudience() == "" || expiresAt == nil || metadata.GetKeyId() == "" {
		return ErrMalformedProductionWire
	}
	if err := expiresAt.CheckValid(); err != nil {
		return ErrMalformedProductionWire
	}
	return nil
}

func validateMetadataForSigning(metadata *agentv1.SignatureMetadata, now time.Time) error {
	if err := validateMetadataShape(metadata); err != nil {
		return err
	}
	expiresAt := metadata.GetExpiresAt().AsTime()
	if !expiresAt.After(now) {
		return ErrSignatureExpired
	}
	if expiresAt.After(now.Add(DefaultProductionMaxLifetime)) {
		return ErrMalformedProductionWire
	}
	return nil
}

func validateMetadata(metadata *agentv1.SignatureMetadata, trust ProductionTrustStore) error {
	if err := validateMetadataShape(metadata); err != nil {
		return err
	}
	maxLifetime := trust.MaxLifetime
	if maxLifetime == 0 {
		maxLifetime = DefaultProductionMaxLifetime
	}
	if maxLifetime < 0 || maxLifetime > DefaultProductionMaxLifetime ||
		trust.ClockSkew < 0 || trust.ClockSkew > MaximumProductionClockSkew ||
		trust.ClockSkew > maxLifetime {
		return ErrMalformedProductionWire
	}
	now := time.Now()
	if trust.Now != nil {
		now = trust.Now()
	}
	if !metadata.GetExpiresAt().AsTime().After(now.Add(-trust.ClockSkew)) {
		return ErrSignatureExpired
	}
	if metadata.GetExpiresAt().AsTime().After(now.Add(maxLifetime + trust.ClockSkew)) {
		return ErrMalformedProductionWire
	}
	if metadata.GetAudience() != trust.Audience {
		return ErrAudienceMismatch
	}
	return nil
}

func resolveProductionKey(packet *agentv1.BusPacket, keyID string, trust ProductionTrustStore) (*ecdsa.PublicKey, error) {
	tenant, sender := packet.GetIdentity().GetTenantId(), packet.GetSenderId()
	if tenant != trust.Tenant || sender != trust.Sender {
		return nil, ErrUnknownKeyID
	}
	if trust.ResolveKey != nil {
		key, err := trust.ResolveKey(tenant, sender, keyID)
		if err != nil || key == nil {
			return nil, ErrUnknownKeyID
		}
		if !validProductionPublicKey(key) {
			return nil, ErrUnsupportedProductionKey
		}
		return key, nil
	}
	key := trust.PublicKeys[keyID]
	if key == nil {
		return nil, ErrUnknownKeyID
	}
	if !validProductionPublicKey(key) {
		return nil, ErrUnsupportedProductionKey
	}
	return key, nil
}

func validProductionPublicKey(key *ecdsa.PublicKey) bool {
	return key != nil && key.Curve == elliptic.P256() && key.X != nil && key.Y != nil &&
		key.Curve.IsOnCurve(key.X, key.Y)
}
